circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,413 @@
+	}
+	if config.Timeout <= 0 {
+		config.Timeout = DefaultConfig.Timeout
+	}
+	if config.SuccessThreshold <= 0 {
+		config.SuccessThreshold = DefaultConfig.SuccessThreshold
+	}
+	if config.HalfOpenMaxConcurrent <= 0 {
+		config.HalfOpenMaxConcurrent = DefaultConfig.HalfOpenMaxConcurrent
+	}
+	if config.IsFailure == nil {
+		config.IsFailure = DefaultConfig.IsFailure
+	}
+	if config.OnOpen == nil {
+		config.OnOpen = DefaultConfig.OnOpen
+	}
+	if config.OnHalfOpen == nil {
+		config.OnHalfOpen = DefaultConfig.OnHalfOpen
+	}
+	if config.OnClose == nil {
+		config.OnClose = DefaultConfig.OnClose
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	now := time.Now()
+
+	return &CircuitBreaker{
+		failureThreshold:  config.FailureThreshold,
+		timeout:           config.Timeout,
+		successThreshold:  config.SuccessThreshold,
+		state:             StateClosed,
+		ctx:               ctx,
+		cancel:            cancel,
+		config:            config,
+		now:               time.Now,
+		halfOpenSemaphore: make(chan struct{}, config.HalfOpenMaxConcurrent),
+		lastStateChange:   now,
+		totalRequests:     0,
+		rejectedRequests:  0,
+	}
+}
+
+// Stop cancels the circuit breaker and releases resources
+func (cb *CircuitBreaker) Stop() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.openTimer != nil {
+		cb.openTimer.Stop()
+	}
+	cb.cancel()
+}
+
+// GetState returns the current state of the circuit breaker
+func (cb *CircuitBreaker) GetState() State {
+	cb.mutex.RLock()
+	defer cb.mutex.RUnlock()
+	return cb.state
+}
+
+// IsOpen returns true if the circuit is open
+func (cb *CircuitBreaker) IsOpen() bool {
+	return cb.GetState() == StateOpen
+}
+
+// Reset resets the circuit breaker to its initial closed state
+func (cb *CircuitBreaker) Reset() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	// Reset counters
+	atomic.StoreInt64(&cb.failureCount, 0)
+	atomic.StoreInt64(&cb.successCount, 0)
+
+	// Reset state
+	cb.state = StateClosed
+	cb.lastStateChange = cb.now()
+
+	// Cancel any pending state transitions
+	if cb.openTimer != nil {
+		cb.openTimer.Stop()
+	}
+}
+
+// ForceOpen forcibly opens the circuit regardless of failure count
+func (cb *CircuitBreaker) ForceOpen() {
+	cb.transitionToOpen()
+}
+
+// ForceClose forcibly closes the circuit regardless of current state
+func (cb *CircuitBreaker) ForceClose() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	cb.state = StateClosed
+	cb.lastStateChange = cb.now()
+	atomic.StoreInt64(&cb.failureCount, 0)
+	atomic.StoreInt64(&cb.successCount, 0)
+
+	if cb.openTimer != nil {
+		cb.openTimer.Stop()
+	}
+}
+
+// SetTimeout updates the timeout duration
+func (cb *CircuitBreaker) SetTimeout(timeout time.Duration) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	cb.timeout = timeout
+}
+
+// transitionToOpen changes state to open and schedules transition to half-open
+func (cb *CircuitBreaker) transitionToOpen() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state != StateOpen {
```

Comment:
this could be `if cb.state == StateOpen { return}` to get rid of 1 level of indentation